libs/checkpoint/langgraph/store/base.py

```diff
@@ -0,0 +1,191 @@
+"""Base classes and types for persistent key-value stores.
+
+Stores enable persistence and memory that can be shared across threads,
+scoped to user IDs, assistant IDs, or other arbitrary namespaces.
+"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Iterable, NamedTuple, Optional, Union
+
+
+@dataclass
+class Item:
+    """Represents a stored item with metadata."""
+
+    value: dict[str, Any]
+    """The stored data."""
+    scores: dict[str, float]
+    """Relevance scores for the item."""
+    id: str
+    """Unique identifier within the namespace."""
+    namespace: tuple[str, ...]
+    """Hierarchical path for organizing items."""
+    created_at: datetime
+    """Timestamp of item creation."""
+    updated_at: datetime
+    """Timestamp of last update."""
+    last_accessed_at: datetime
+    """Timestamp of last access."""
+
+
+class GetOp(NamedTuple):
+    """Operation to retrieve an item by namespace and ID."""
+
+    namespace: tuple[str, ...]
+    """Hierarchical path for the item."""
+    id: str
+    """Unique identifier within the namespace."""
+
+
+class SearchOp(NamedTuple):
+    """Operation to search for items within a namespace prefix."""
+
+    namespace_prefix: tuple[str, ...]
+    """Hierarchical path prefix to search within."""
+    filter: Optional[dict[str, Any]] = None
+    """Key-value pairs to filter results."""
+    limit: int = 10
+    """Maximum number of items to return."""
+    offset: int = 0
+    """Number of items to skip before returning results."""
+
+
+class PutOp(NamedTuple):
+    """Operation to store or update an item."""
+
+    namespace: tuple[str, ...]
+    """Hierarchical path for the item."""
+    id: str
+    """Unique identifier within the namespace."""
+    value: Optional[dict[str, Any]]
+    """Data to be stored, or None to delete."""
+
+
+Op = Union[GetOp, SearchOp, PutOp]
+Result = Union[Item, list[Item], None]
+
+
+class BaseStore(ABC):
+    """Abstract base class for key-value stores."""
+
+    __slots__ = ("__weakref__",)
+
+    @abstractmethod
+    def batch(self, ops: Iterable[Op]) -> list[Result]:
+        """Execute a batch of operations synchronously."""
+
+    @abstractmethod
+    async def abatch(self, ops: Iterable[Op]) -> list[Result]:
```

Comment:
Iterable input, list otuput type.

Did we want to do `Sequence` input instead? I don't see a need to support generators here